Stop storing predefined exceptions in the cu_required_global field of compilation unit descriptiors

shindere opened the pull request:
This is split out of #12031 as suggested and after offline discussion with
@Octachron.

I think a lot has already been explained in the discussion there, but
basically the aim ofthecurrent PR is to stop storing names of predefined
exceptionsin  the cu_required_globals filed of compilation units, which
should store only thenames of compilation units to be loaded before the one
described because their initialisation ocde needs to be executed before
theinitialisation codeof the currently described compilation unit.

shindere committed:
Do not add predefined identifiers to the set of required globals

This needs to be done both in lambda/translmod and in bytecomp/symtable.

shindere committed:
Compiler bootstrap

At this stage it is guaranteed that the compilation units written by
boot/ocamlc will not contain any predefined symbols in their
"required globals" section.

It will thus become possible to stop filtering out those predefined
symbols but the intermediate bootstrap is required.

lthls commented on bytecomp/symtable.ml:
```diff
@@ -348,9 +348,10 @@ let defined_globals patchlist =
   List.fold_left (fun accu rel ->
       match rel with
-      | (Reloc_getglobal id, _pos) -> id :: accu
+      | (Reloc_getglobal id, _pos) when (not_predef id) -> id :: accu
```
I have a slight preference for the following style:
```ocaml
| (Reloc_getglobal id, _pos) -> if Ident.is_predef id then accu else id :: accu
```
If you prefer your version, I'd still suggest to name your auxiliary function `is_comp_unit` instead of `not_predef`.

lthls commented on lambda/translmod.ml:
```diff
@@ -763,11 +763,12 @@ let _ =
     Lambda.iter_head_constructor scan lam;
     match lam with
-      Lprim ((Pgetglobal id | Psetglobal id), _, _) ->
+      Lprim ((Pgetglobal id | Psetglobal id), _, _) when (not_predef id) ->
```
Same remark as in `symtable.ml`

lthls reviewed (commented):
In addition to some style suggestions, I'd like to understand why the second bootstrap is necessary.
Apart from that I'm in favour of the change and the code looks correct.

shindere commented:
Many thanks for your prompt review, Vincent.

Vincent Laviron (2023/03/17 06:42 -0700):
>    @lthls commented on this pull request.
>
>    In addition to some style suggestions, I'd like to understand why the
>    second bootstrap is necessary.

I think you are right it is not strictly necessary so I got rid of it.
It just means that boot/ocamlc will continue to do some useless
filteringfor sometime but it's certainly something we canlivewith,
especially given that #12031 which will hopefully also get merged soon
(when reworked) willinclude have to include a bootstrap, which will make
htis useless filteringgo away anyway. So yes, let's save one bootstrap
here.

>    Apart from that I'm in favour of the change and the code looks correct.
>      __________________________________________________________________
>
>    In [1]bytecomp/symtable.ml:
> >    List.fold_left (fun accu rel ->
>        match rel with
> -      | (Reloc_getglobal id, _pos) -> id :: accu
> +      | (Reloc_getglobal id, _pos) when (not_predef id) -> id :: accu
>
>    I have a slight preference for the following style:
>
> | (Reloc_getglobal id, _pos) -> if Ident.is_predef id then accu else id :: accu
>
>    If you prefer your version, I'd still suggest to name your auxiliary
>    function is_comp_unit instead of not_predef.

I used your second proposal, thanks. I thknk it provides a nicer basis
for #12031 which will rewrite thispart of thecode anyway.


>      __________________________________________________________________
>
>    In [2]lambda/translmod.ml:
> >    let globals = ref Ident.Set.empty in
>    let rec scan lam =
>      Lambda.iter_head_constructor scan lam;
>      match lam with
> -      Lprim ((Pgetglobal id | Psetglobal id), _, _) ->
> +      Lprim ((Pgetglobal id | Psetglobal id), _, _) when (not_predef id) ->
>
>    Same remark as in symtable.ml

Also fixed. Thanks.

I also reduced the number of reviewers in the Changes entry. I
startedwiththe list in #12031 but since people who participated there
will be credited there anyway, it felt more accurate tome to have a
smaller list here.

shindere committed:
Stop filtering out predefined identifiers from cu_required_globals

Now that it is guaranteed that the cu_required_globalbs field of
compilation units does not contain any predefined symbol, it becomes
possible to get rid of the code that was used to filter them out.

shindere commented:
Vincent Laviron (2023/03/17 07:21 -0700):
> @lthls approved this pull request.

Thanks a lot, Vincent.

This has now been merged.
